gaps/gap-25_golem_certificates/gap-25_golem_certificates.md

```diff
@@ -0,0 +1,194 @@
+
+#### Creating a signature
+
+In order to cryptographically sign a certificate the following steps need to be taken:
+
+1. Create a binary representation of the `certificate` property by serializing the content via the above mentioned `JSON Canonicalization Scheme`
+2. Feed the binary data created in step 1. into a hashing algorithm that is compatible with the chose encryption to obtain the fingerprint of the certificate
+3. Encrypt the fingerprint obtained in step 2. via the encryption algorithm supported by the private key of the signer. The private key must be the one connected to the public key specified in the signer's certificate. This encrypted fingerprint is the signature.
+4. Create the `signature` property in the JSON certificate and add the following details
+  - name of the hash and encryption algorithm used to create the signature into the `algorithm` property
+  - binary data of the signature into `value` property
+  - the `signer` property of the `signature` should contain the signing certificate or the string `self` in case of self signature
+
+#### Verifying a signature
+
+1. Verify that the signing certificate contained in `signature.signer` is allowed to sign certificates (it's `keyUsage` property is either set to the string `all` or the list contains `signCertificate`).
+2. Create a binary representation of the `certificate` property by serializing the content via the above mentioned `JSON Canonicalization Scheme`
+3. Feed the binary data created in step 1. into the hashing algorithm defined in the `signature.algorithm.hash` property of the certificate to obtain the fingerprint of the certificate
+4. Decrypt the signature stored in `signature.value` using the public key of the signer (the `certificate.publicKey` property of the signing certificate) and the encryption algorithm defined in `signature.algorithm.encryption`. The result of this step will be the signed fingerprint.
+5. Verify that the fingerprint obtained in step 2. and step 3. are the same.
+
+### Verifying a certificate chain
+
+A signed certificate is only valid if the cryptographic signature is valid and the following constraints are fulfilled:
+- The certificate cannot have a validity period that is not fully contained in the signer's validity period. The certificates can have the same `notBefore` and `notAfter` values but the signer cannot sign any certificate that would 'outlive' it.
```

Comment:
We also need to check if current datetime is contained inside validity period (expired / not valid yet)

Reply:
Ah yes, thank you.